File: BehavioralPattern/TemplateMethod/TemplateMethod.cc
#include <iostream>
using namespace std;
//制作饮料
class Beverage
{
public:
    //	煮⽔
    void BoilWater()
    {
        cout << "烧开⽔" << endl;
    }
    //	冲泡		特殊接⼝
    virtual void Brew() = 0;
    //	倒⼊杯中
    void PourInCup()
    {
        cout << "倒⼊杯中" << endl;
    }
    //	添加料 特殊接⼝
    virtual void AddCondiments() = 0;
    //	hook	⽅法，决定某些算法步骤是否挂钩在算法中
    virtual bool CustomWantsCondiments()
    {
        return true;
    }
    //	模板⽅法
    void PrepareBeverage()
    {
        BoilWater();
        Brew();
        PourInCup();
        if (CustomWantsCondiments())
        {
            AddCondiments();
        }
    }
};

class Coffee : public Beverage
{
public:
    virtual void Brew()
    {
        cout << "冲泡咖啡⾖" << endl;
    }
    virtual void AddCondiments()
    {
        cout << "添加糖和⽜奶	" << endl;
    }
};

class Tea : public Beverage
{
public:
    virtual void Brew()
    {
        cout << "冲泡茶叶	" << endl;
    }
    virtual void AddCondiments()
    {
        cout << "添加柠檬	" << endl;
    }
};
int main(void)
{
    Beverage *pTea = new Tea;
    pTea->PrepareBeverage();
    delete pTea;
    cout << "---------------------" << endl;
    Beverage *pCoffee = new Coffee;
    pCoffee->PrepareBeverage();
    delete pCoffee;
    return 0;
}

File: BehavioralPattern/TemplateMethod/TemplateMethod.go
package main

import "fmt"

type IBeverage interface {
	// 煮⽔
	BoilWater()
	//	冲泡
	Brew()
	// 倒⼊杯中
	PourInCup()
	//	添加料 特殊接⼝
	AddCondiments()
}
type Beverage struct {
}

func (b *Beverage) BoilWater() {
	fmt.Println("煮水")
}
func (b *Beverage) Brew() {

}
func (b *Beverage) PourInCup() {
	fmt.Println("倒入杯中")
}

//	添加料 特殊接⼝,留给子类实现
func (b *Beverage) AddCondiments() {

}
func PrepareBeverage(i IBeverage) {
	i.BoilWater()
	i.AddCondiments()
	i.Brew()
	i.PourInCup()
}

type Coffee struct {
	*Beverage
}

func (b *Coffee) Brew() {
	fmt.Println("冲泡咖啡")
}
func (c *Coffee) AddCondiments() {
	fmt.Println("加入咖啡")
}

type Tea struct {
	*Beverage
}

func (b *Tea) Brew() {
	fmt.Println("冲泡茶叶")
}
func (c *Tea) AddCondiments() {
	fmt.Println("加入茶叶")
}

func main() {
	var b Beverage
	var i IBeverage = &Coffee{
		Beverage: &b,
	}
	PrepareBeverage(i)
	var i2 IBeverage = &Tea{
		Beverage: &b,
	}
	PrepareBeverage(i2)
}
